enhancement(phone-input) implement type to search and bugfixes

qqingvarqq opened the pull request:
Resolves #1462
#### Description
PhoneInput:
1. Implement filter functionality.
2. Propagate down missing property (maxDropdownHeight).
3. Enable adjusting of dropdown height when we have a small number of countries (mean when countries are filtering).
4. ScrolltoIndex now works correctly when we set different `maxDropdownHeight` property.
5. Add types (TS, Flow) (Also update incorrect ts types).
6. Update screener and e2e tests.
7. Add examples to docs for phone input with filter.
----
Note:
1. To phone input component I add new property `enableFiltering` to show or not show filtering, so existing users wouldn't see any changes until they do not enable filtering explicitly.
(Potentially this PR does not introduce any breaking changes.)
2. I don't know if I add correct styles, because you don't provide any sketches. let me know if need change styles.
3. Changes to ` documentation-site/components/yard/override.tsx,  documentation-site/examples/popover/placements.js` added automatically by husky prehook (It's just prettier changes but if you need I can revert them)
#### Scope

- [X] Patch: Bug Fix
- [X] Minor: New Feature
- [ ] Major: Breaking Change

qqingvarqq commented:
To check how it works,  check netlify build:
https://5d4ae483c012ba000b46d956--baseui.netlify.com/components/phone-input/#basic-with-enabled-dropdown-filter

sandgraham commented on src/phone-input/country-select-dropdown.js:
```diff
@@ -17,11 +18,38 @@ import {
+  return Math.min(actualHeight, height) + 'px';
+}
+
+function calcualteInitialScrollIndex(
```
typo: `calculateInitialScrollIndex`

gergelyke commented:
this looks amazing @qqingvarqq ! can we enable better keyboard navigation for the component? it seems like I cannot navigate into the search field using the keyboard

nadiia commented on src/phone-input/country-select-dropdown.js:
```diff
@@ -60,59 +101,80 @@ function CountrySelectDropdown(
+        $height={calculateDropDownHeight(
+          maxDropdownHeight,
+          children.length,
+          LIST_ROW_HEIGHT,
```
Not sure about using a constant here for the row height. Is there a case when the height is larger/smaller due to text wrapping or font change?

qqingvarqq replied:
We need this constant here to calculate the correct height of dropdown for example when only 2 items satisfy filter condition. I don't saw these cases but maybe @sandgraham know more?

nadiia commented on src/phone-input/country-select-dropdown.js:
```diff
@@ -60,59 +101,80 @@ function CountrySelectDropdown(
+                        style={style}
+                        {...rest}
+                        {...listItemProps}
+                        data-e2e="country-select-list-item"
```
This attribute should be added through overrides for the test example only.

qqingvarqq commented:
I will add keyboard navigation on the weekend
I don't use keyboard navigation often. So I guess correct behavior would be:
1. When countries dropdown will open, pressing `tab` will shift focus to the filter input and if we press `tab` again dropdown menu will close and focus would be on a phone input part?
2. Pressing escape, when filter input is focused, will close dropdown?
@nadiia @gergelyke @sandgraham  am I correct?

--------
One thought Since country select dropdown is implemented on top of the select component. It will not easy to access the internal state of the select component and will complicate things a lot when I implement this keyboard navigation (at least now I can't see the really good solution).
And there are some other problems I noticed and would solve. (it's affecting this version and master branch version).
First, it was a keyboard navigation issue with the dropdown when you just open the select menu and press `enter` it changes country to the first country in a list, which is not good.
Second, when you press `arrowUp` and `arrowDown` the list wouldn't scroll. But in regular (not virtualized) select it works. To fix this we can use arrowKeyStepper [demo](http://bvaughn.github.io/react-virtualized/#/components/ArrowKeyStepper) from `react-virtualized` and remove StateFullMenu witch used by Select component.

So to make it simpler I think about replace select with own implementation.

Any thoughts, suggestions?

houmark commented:
Great stuff. I tested this in the Netlify build. It seems like searching for the dial code is not producing results. That means I have to search for the country name to get a result. Should it not include the dial code also?

houmark commented:
Did this get stuck? @qqingvarqq will you be able to work on this or should I consider taking on the changes requested to get this released?

qqingvarqq commented:
@houmark  Nope hope I will update this PR,  up to Thursday maybe sooner.

qqingvarqq commented:
not ready for review !!!

qqingvarqq commented:
WIP

TODO

- Update all requested changes
- Update tests

@gergelyke @sandgraham @nadiia 
I improved keyboard navigation and fixed those issues which you are pointed, so it's the almost final version. For some reason, netlify build was failed, (`9:24:59 PM: Started saving pip cache` and killed because Command did not finish within the time limit) so you probably can't easily see changes.
BTW,  About @houmark  question
 `It seems like searching for the dial-code is not producing results.` 
For me, it's a very strange case, when someone knows phone dial-code but don't know in which country he remains, so why he would start typing dial-code in the dropdown menu?
What your opinion on it?

houmark commented:
> BTW, About @houmark question
It seems like searching for the dial-code is not producing results.
For me, it's a very strange case, when someone knows phone dial-code but don't know in which country he remains, so why he would start typing dial-code in the dropdown menu?

Many times when you type in a phone number in a UI, it's your own phone number, so for me it's faster to type 2 or 3 digits than a country name that is similarly named to others. And normally one would expect that visible values in the UI is also searchable, so when you open the list and see both the countries and dial codes, then you may expect you can search for the dial code. Users think differently, making it search for both values solves any use case.

Thanks for your great work on this!

qqingvarqq commented:
ready for review 🎉
Changes:

- add keyboard accessibility
- small bugfixes
- simplified dom structure of dropdown
- updated tests and docs
- add tests for keyboard accessibility
- add requested changes
- add a possibility to use dial code in the search menu

houmark commented:
@qqingvarqq Happy to see it's searching nicely for the dial code now also.

By testing the Netlify build (which I believe is up to date now) I noticed the following UX issues:

1) When you have found your result, there's no way to use the keyboard to select the country (arrow down does not work). Additionally, it could maybe auto-select the first result, which would also mean that one result would be select and then hitting tab would change to that country? The only way now is using the mouse to click the result.

2) If I have a search and click the flag to effectively close the list, and re-open, it will then start over and forget my search. Maybe that should be persisted?

Besides that, this is really close to the perfect solution for me and it looks great also.

qqingvarqq commented:
@houmark 

>  it looks great also

Thanks)

> When you have searched and got your result in the list, there's no way to use the keyboard to select the country

I believe you miss something, pressing `enter` would select a country and `arrow up/down` works with focused input, also when you filter countries it autoselects the first result.
I checked it works:

![info](https://user-images.githubusercontent.com/13419634/63635516-320d2280-c66c-11e9-849b-7b1aa32b2355.gif)

> which would also mean that one result would be select and then hitting tab would change to that country

 If one result left, we would automatically select this result, so just pressing enter selects him, there no need in additional tabbing.

----

> If I have a search and click the flag to effectively close the list, and re-open, it will then start over and forget my search. Maybe that should be persisted?

I don't really like this idea of saving filter mask, because I believe when you open this dropdown it's important to get all countries and if need type filter mask. For example what if the user types something that ends up in zero results and closes the dropdown, and when he opened it again he got the message `No results`.

houmark commented:
@qqingvarqq It looks like I may have hit an old version of the Netlify demo or something because as you can see in demo, it did not work when I tested this, but it does now.

Persisting the filter mask is very minor to me, and while a solution could be that no results would result a reset upon re-open while keeping results if there's > 0 results, this is nitpicking and just my personal preference.

This works great and not considering coding conventions, I can give this my +1 from testing and hope to see this merged soon!

qqingvarqq commented on src/select/types.js:
```diff
@@ -229,6 +230,9 @@ export type DropdownPropsT = {
   value: ValueT,
   valueKey: string,
   width: ?number,
+  selectKeyDownHandler: (event: KeyboardEvent) => void,
```
Yes, I can do it.

gergelyke commented:
@qqingvarqq are you planning to work on this one?

qqingvarqq commented:
@gergelyke  yes, I'll update my PR soon (hopefully next week), was busy for the last few weeks.

qqingvarqq commented:
I will be little busy on the next weeks, so don't know how soon I can update this Pr. I will split this PR on a few fewer PRs, it will allow to review changes easier, and all changes can go to master branch more quickly.

gergelyke commented:
> I will be little busy on the next weeks, so don't know how soon I can update this Pr. I will split this PR on a few fewer PRs, it will allow to review changes easier, and all changes can go to master branch more quickly.
> Part 1 -> #1970

sounds great, thank you - I'll close this PR in favor of the new follow-up PRs
